### PR TITLE
add op_checksigadd

### PR DESCRIPTION
**Pull request by harsh-ps-2003**
- [ ] issue #
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/shinigami/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

Adds opcode checksigadd!
Reference - https://en.bitcoin.it/wiki/BIP_0342

### CODE REVIEW

**Comment by harsh-ps-2003** [comment]
@b-j-roberts can you take a look why am I getting i64 overflow :(

File: src/opcodes/tests/test_crypto.cairo [hunk 1]
```diff
@@ -382,6 +382,18 @@ fn test_op_sha1_14_double_sha1() {
     utils::check_expected_dstack(ref engine, expected_stack.span());
 }
 
+#[test]
+fn test_op_checksigadd_valid() {
+    let script_sig =
+        "OP_DATA_71 0x3044022008f4f37e2d8f74e18c1b8fde2374d5f28402fb8ab7fd1cc5b786aa40851a70cb02201f40afd1627798ee8529095ca4b205498032315240ac322c9d8ff0f205a93a5801 OP_DATA_33 0x024aeaf55040fa16de37303d13ca1dde85f4ca9baa36e2963a27a1c0c1165fe2b1";
+    let script_pubkey = "OP_1 OP_SWAP OP_SWAP OP_CHECKSIGADD";
```
Comment (j1mbo64) [inline src/opcodes/tests/test_crypto.cairo:389]: You do OP_SWAP twice, so the stack is again [SIG, PK, N] and after you try to make a number with the public key which result in an overflow. Can you try to remove one OP_SWAP ?

File: src/opcodes/crypto.cairo [hunk 2]
```diff
@@ -238,3 +239,48 @@ pub fn opcode_sha1(ref engine: Engine) -> Result<(), felt252> {
     engine.dstack.push_byte_array(h);
     return Result::Ok(());
 }
+
+pub fn opcode_checksigadd(ref engine: Engine) -> Result<(), felt252> {
+    let pk_bytes = engine.dstack.pop_byte_array()?;
+    let n = engine.dstack.pop_int()?;
+
+    if n > 0x7FFFFFFF || n < -0x80000000 {
```
Comment (j1mbo64) [inline src/opcodes/crypto.cairo:247]: This test is already done at line 245 in `pop_int`

File: src/opcodes/crypto.cairo [hunk 3]
```diff
@@ -238,3 +239,48 @@ pub fn opcode_sha1(ref engine: Engine) -> Result<(), felt252> {
...
+
+    let sig_bytes = engine.dstack.pop_byte_array()?;
+
+    if pk_bytes.len() == 0 {
+        return Result::Err('EmptyPublicKey');
+    }
+
+    let mut is_valid = false;
+
+    if pk_bytes.len() == 32 {
```
Comment (j1mbo64) [inline src/opcodes/crypto.cairo:259]: I think this test can be removed, `verify` result as an error if the key is not in the good format. You're also pushing 33 bytes public key so this test is always false

File: src/opcodes/crypto.cairo [hunk 4]
```diff
@@ -238,3 +239,48 @@ pub fn opcode_sha1(ref engine: Engine) -> Result<(), felt252> {
...
+            let mut sig_verifier = BaseSigVerifierTrait::new(ref engine, @sig_bytes, @pk_bytes)?;
+            is_valid = sig_verifier.verify(ref engine);
+            if !is_valid {
+                return Result::Err('InvalidSignature');
+            }
+        }
+    } else {
+        // Unknown public key type
+        if sig_bytes.len() > 0 {
+            is_valid = true;
```
Comment (j1mbo64) [inline src/opcodes/crypto.cairo:270]: Your test `test_checksigadd_valid` pass because you execute this

File: src/opcodes/crypto.cairo [hunk 5]
```diff
@@ -238,3 +239,48 @@ pub fn opcode_sha1(ref engine: Engine) -> Result<(), felt252> {
...
+    let n = engine.dstack.pop_int()?;
+
+    if n > 0x7FFFFFFF || n < -0x80000000 {
+        return Result::Err('InvalidScriptNum');
+    }
+
+    let sig_bytes = engine.dstack.pop_byte_array()?;
+
+    if pk_bytes.len() == 0 {
+        return Result::Err('EmptyPublicKey');
```
Comment (j1mbo64) [inline src/opcodes/crypto.cairo:254]: You should create an error in `Errors.cairo`

**Review by j1mbo64 — CHANGES_REQUESTED** [review]

**Comment by harsh-ps-2003** [comment]
@j1mbo64 sorry! I dont have the bandwidth to fix the tests! Thank you for the review! Maybe you can reopen the PR and take up the work further if you want!

**Pull request closed without merging**